### PR TITLE
Arg resolver cleanup

### PR DESCRIPTION
**Pull request by bdistin**
### Proposed Semver Increment Bump: [MAJOR/MINOR/PATCH]
patch
### Changes Proposed in this Pull Request (List new items in CHANGELOG.MD)

-
-
-

### (If Applicable) What Issue does it fix?
 Fixes... uncached users/members from failing checks (fetches them if not cached)

### CODE REVIEW

**Review by UnseenFaith — DISMISSED** [review]
lgtm

File: classes/argResolver.js [hunk 1]
```diff
@@ -12,90 +21,70 @@ module.exports = class ArgResolver {
...
-
   async mention(arg, currentUsage, possible, repeat) {
     return this.user(arg, currentUsage, possible, repeat);
   }
 
+  async user(arg, currentUsage, possible, repeat) {
+    // eslint-disable-next-line no-nested-ternary
+    const user = regex.userOrMember.test(arg) && this.client.user.bot ?
+      await this.client.fetchUser(regex.userOrMember.exec(arg)[1]).catch(() => null) :
+      (regex.userOrMember.test(arg) ? this.client.users.get(regex.userOrMember.exec(arg)[1]) : undefined);
```
Comment (kyranet) [inline classes/argResolver.js]: This regex is being processed twice.

File: classes/Resolver.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  async boolean(bool) {
+    if (bool instanceof Boolean) return bool;
+    if ([1, "true", "+", "t"].includes(bool)) return true;
+    if ([0, "false", "-", "f"].includes(bool)) return false;
+    return null;
+  }
+
+  async string(string) {
+    if (string instanceof String) return true;
```
Comment (kyranet) [inline classes/Resolver.js]: Resolver#string is supposed to return a string, not boolean.

File: classes/Resolver.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    if (role instanceof Discord.Channel) return role;
+    if (role instanceof Discord.Role) return role;
+    if (role instanceof String) return regex.role.test(role) ? guild.roles.get(regex.role.exec(role)[1]) : null;
+    return null;
+  }
+
+  async boolean(bool) {
+    if (bool instanceof Boolean) return bool;
+    if ([1, "true", "+", "t"].includes(bool)) return true;
+    if ([0, "false", "-", "f"].includes(bool)) return false;
```
Comment (kyranet) [inline classes/Resolver.js]: `[1].includes("1");` returns false. Both 1 and 0 must be instringified. Same for the case in false.

Comment (kyranet) [inline classes/Resolver.js]: Same for the case in false.

File: classes/Resolver.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      return regex.userOrMember.test(user) ?
+      this.client.user.bot ? this.client.fetchUser(regex.userOrMember.exec(user)[1]).catch(() => null) : this.client.users.get(regex.userOrMember.exec(user)[1]) :
+      null;
+    }
+    return null;
+  }
+
+  async member(member, guild) {
+    if (member instanceof Discord.Member) return member;
+    if (member instanceof String) return regex.userOrMember.test(member) ? guild.fetchMember(regex.userOrMember.exec(member)[1]).catch(() => null) : null;
```
Comment (kyranet) [inline classes/Resolver.js]: Same here.

File: classes/Resolver.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    return null;
+  }
+
+  async channel(channel) {
+    if (channel instanceof Discord.Channel) return channel;
+    if (channel instanceof String) return regex.channel.test(channel) ? this.client.channels.get(regex.channel.exec(channel)[1]) : null;
```
Comment (kyranet) [inline classes/Resolver.js]: Same here.

File: classes/Resolver.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    return null;
+  }
+
+  async guild(guild) {
+    if (guild instanceof Discord.Channel) return guild;
+    if (guild instanceof String) return regex.snowflake.test(guild) ? this.client.guilds.get(guild) : null;
```
Comment (kyranet) [inline classes/Resolver.js]: Here...

File: classes/Resolver.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    return null;
+  }
+
+  async role(role, guild) {
+    if (role instanceof Discord.Channel) return role;
+    if (role instanceof Discord.Role) return role;
+    if (role instanceof String) return regex.role.test(role) ? guild.roles.get(regex.role.exec(role)[1]) : null;
```
Comment (kyranet) [inline classes/Resolver.js]: And here.

File: classes/Resolver.js [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+  async msg(message, channel) {
+    if (message instanceof Discord.Message) return message;
+    return regex.snowflake.test(message) ? channel.fetchMessage(message).catch(() => null) : undefined;
+  }
+
+  async user(user) {
+    if (user instanceof Discord.User) return user;
+    if (user instanceof Discord.Member) return user.user;
+    if (user instanceof String) {
```
Comment (kyranet) [inline classes/Resolver.js]: member instanceof String will always return false, unless doing a previous `String(member);`. However, `member.constructor.name === "String";` works perfectly.

**Review by kyranet — CHANGES_REQUESTED** [review]
There are some changes that won't work as expected.

**Review by kyranet — DISMISSED** [review]

**Review by kyranet — DISMISSED** [review]
Looks good now.

File: classes/Resolver.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  async channel(channel) {
+    if (channel instanceof Discord.Channel) return channel;
+    if (typeof channel === "string") return regex.channel.test(channel) ? this.client.channels.get(regex.channel.exec(channel)[1]) : null;
+    return null;
+  }
+
+  async guild(guild) {
+    if (guild instanceof Discord.Channel) return guild;
```
Comment (kyranet) [inline classes/Resolver.js]: Shouldn't this be instanceof Discord.Guild?

File: classes/Resolver.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+  async boolean(bool) {
+    if (bool instanceof Boolean) return bool;
+    if (["1", "true", "+", "t", "yes"].includes(String(bool).toLowerCase())) return true;
+    if (["0", "false", "-", "f", "no"].includes(String(bool).toLowerCase())) return false;
+    return null;
+  }
+
+  async string(string) {
+    if (typeof string === "string") return string;
```
Comment (kyranet) [inline classes/Resolver.js]: In both ways, it returns a String. Why not just `return String(string);`.

File: classes/argResolver.js [hunk 11]
```diff
@@ -104,88 +79,51 @@ module.exports = class ArgResolver {
...
-          return null;
-        }
-        throw `${currentUsage.possibles[possible].name} must be shorter than ${currentUsage.possibles[possible].max} characters.`;
-      } else {
-        return arg;
-      }
-    } else {
-      return arg;
+    const min = currentUsage.possibles[possible].min;
+    const max = currentUsage.possibles[possible].max;
```
Comment (kyranet) [inline classes/argResolver.js]: You can use deconstruction here. `const { max } = currentUsage.possibles[possible];`

File: classes/argResolver.js [hunk 12]
```diff
@@ -104,88 +79,51 @@ module.exports = class ArgResolver {
...
   }
 
   async integer(arg, currentUsage, possible, repeat) {
-    arg = Number(arg);
-    if (!Number.isInteger(arg)) {
-      if (currentUsage.type === "optional" && !repeat) {
-        return null;
-      }
+    const min = currentUsage.possibles[possible].min;
+    const max = currentUsage.possibles[possible].max;
```
Comment (kyranet) [inline classes/argResolver.js]: Same here.

File: classes/argResolver.js [hunk 13]
```diff
@@ -1,101 +1,76 @@
...
-  }
-
   async mention(arg, currentUsage, possible, repeat) {
     return this.user(arg, currentUsage, possible, repeat);
   }
 
+  async user(arg, currentUsage, possible, repeat) {
+    const user = await super.user(arg);
+    if (user) return user;
+    if (!currentUsage) return null;
```
Comment (kyranet) [inline classes/argResolver.js]: Isn't this always present?

**Review by kyranet — CHANGES_REQUESTED** [review]
I know all methods are async to keep consistency, but I think it's not needed if we don't use async operations.

**Comment by kyranet** [comment]
To be honest, only the member parser needs async/await, as it fetchs the user and then the member, but the rest returns either a promise or the output, slowing down the parser for the async methods isn't the ideal, when only user and member can return a Promise.

File: classes/argResolver.js (hunk 13 above)
Comment (bdistin) [reply classes/argResolver.js]: It isn't... The parent class calls user internally, but since it is extended, the parent call calls the child class...

Comment (kyranet) [reply classes/argResolver.js]: Well, it's wasn't present on the rest of functions.

**Comment by ghost** [comment]
Good stuff here!

**Review by kyranet — DISMISSED** [review]
LGTM.

**Review by CyberiumShadow — APPROVED** [review]
Tested. LGTM

**Pull request merged**